sycl/doc/extensions/experimental/sycl_ext_oneapi_spec_constant_length_alloca.asciidoc

```diff
@@ -0,0 +1,195 @@
+permission by Khronos.
+
+
+== Contact
+
+To report problems with this extension, please open a new issue at:
+
+https://github.com/intel/llvm/issues
+
+
+== Dependencies
+
+This extension is written against the SYCL 2020 revision 8 specification.  All
+references below to the "core SYCL specification" or to section numbers in the
+SYCL specification refer to that revision.
+
+
+== Status
+
+This is an experimental extension specification, intended to provide early
+access to features and gather community feedback.  Interfaces defined in this
+specification are implemented in {dpcpp}, but they are not finalized and may
+change incompatibly in future versions of {dpcpp} without prior notice.
+*Shipping software products should not rely on APIs defined in this
+specification.*
+
+== Overview
+
+C++ arrays and `std::array` can be used in SYCL code to represent fixed-size
+sequences of objects. However, these constructs have a significant restriction:
+the number of elements must be known at compile time. In host-code context,
+users can make use of dynamic memory allocations, e.g., `std::vector`, but this
+is not the case in SYCL device code.
+
+SYCL specialization constants (SYCL Section 4.9.5.) can be used to represent
+constants whose values can be set dynamically during the execution of a SYCL
+application, but that will not change when a SYCL kernel function is
+invoked. This way, specialization constants could be used to implement SYCL
+private arrays whose size is given during the execution of the SYCL
+application. There is no possible way of implementing this using `std::array`,
+as the size of such container must be known at compile time, so we propose to
+define a new `private_alloca` function whose size is specified using SYCL
+specialization constants.
+
+[NOTE]
+====
+This extension only supports SPIR-V backends for now, as it relies on
+SPIR-V-specific capabilities,
+such as specialization constants.
+====
+
+== Specification
+
+=== Feature test macro
+
+This extension provides a feature-test macro as described in the core SYCL
+specification.  An implementation supporting this extension must predefine the
+macro `SYCL_EXT_ONEAPI_SPEC_CONSTANT_LENGTH_ALLOCA` to one of the values defined
+in the table below. Applications can test for the existence of this macro to
+determine if the implementation supports this feature, or applications can test
+the macro's value to determine which of the extension's features the
+implementation supports.
+
+[%header,cols="1,5"]
+|===
+|Value
+|Description
+
+|1
+|The APIs of this experimental extension are not versioned, so the
+ feature-test macro always has this value.
+|===
+
+=== The `private_alloca` function
+
+[source,c++]
+----
+namespace sycl::ext::oneapi::experimental {
+template <typename ElementType, auto &SpecName,
+          access::decorated DecorateAddress>
+private_ptr<ElementType, DecorateAddress>
+private_alloca(kernel_handler &h);
```

Comment:
I think it's problematic that this function is called `alloca` but doesn't really look or behave like `alloca`.  Compare:

```c++
void* bytes = alloca(N);
```
and
```c++
private_ptr<float> ptr = sycl::private_alloca<float, N>(kh);
```

It seems really unnatural to me that this function doesn't accept a size.  Could we use a design like this instead:
```c++
void* bytes = sycl::alloca(kh.get_specialization_constant_something<N>);
``` 

(In hindsight, having `get_specialization_constant` return a value directly may have been a mistake.  The intent of this `get_specialization_constant_something` placeholder is to return a representation of the specialization constant that's associated with a specific kernel handler, rather than the specialization constant's value.)

Reply:
> For context: Our initial design had a different name for the function (allocate_private_memory). We decided to rename as we got comments suggesting people would understand the lifetime of the allocated memory better that way, as reading ...alloca would suggest that was stack allocated memory.

I agree that aligning with `alloca` will help people to understand what the function does.  But I think that alignment comes with a cost, and that we should strive to match the interface where possible.  Otherwise, we've increased the cognitive burden for programmers: people who already know how `alloca` works will read the code and probably think `kh` is a size, and learning the new behavior requires them to learn a new special-case.

> Returning a multi_ptr instead of void * was decided as not only it follows conventions set by other extensions, e.g., [sycl_ext_oneapi_local_memory](https://github.com/intel/llvm/blob/76a8d165a0e657231f3e80a227ad9ef672a650ce/sycl/doc/extensions/supported/sycl_ext_oneapi_local_memory.asciidoc), but also follows sycl::malloc<T>(size_t count, ...) and adds a layer of abstraction. Users will end up doing private_alloca(sizeof(ElementType) * ...), so we are saving that bit.

`sycl::malloc<T>` exists, but so does `sycl::malloc`.  Supporting both options might be worth considering.

> The decision on passing specialization_id as a template argument and kernel_handler & comes from the current specialization constant semantics in the spec. specialization_id, as you know, is just the specialization constant ID and there is no way of getting the specialization constant itself. What you propose, getting a specialization constant reference or similar, might be a better option, but I don't think we want to introduce such a big change to specialization constants in this small extension.

I have a slightly different perspective here, which is that extensions like these are the perfect place to think about what we could/should change in a future version of SYCL.  If we think that a new way of passing specialization constants to functions is desirable, we could prototype it here and move it to a separate extension later.

If you're worried about making changes to `kernel_handler` here, I realized you don't have to.  It might even be cleaner to define a new class, used like this:
```c++
auto size = sycl::spec_constant<Size>(kh);
auto ptr = sycl::alloca(size);
```
...which should allow the element type and specialization constant templates to be deduced.  I think something like this would be much cleaner than a load of required template arguments, especially in the aligned case.

Again, I don't mean to block the initial implementation and experimentation work here, so adding an issue about this would satisfy me.

Tagging @gmlueck for awareness, since I know he has done a lot of work on specialization constants.

Reply:
I added an issue for the `specialization_constant` interface. I'll address the `void *` returning interface and aligned version of the function tomorrow

Reply:
My thinking is that it would be nice not to tie this function to `multi_ptr`, because lots of codes use only raw pointers.

Am I right in thinking that `private_ptr` can be implicitly cast to the underlying pointer type?  That would be enough to enable the usage pattern I have in mind.

Reply:
> Am I right in thinking that private_ptr can be implicitly cast to the underlying pointer type?

you are (like any `multi_ptr`) but it is deprecated in 2020

Reply:
`void *ptr = private_alloca<bytes>(kh).get_raw()` (regardless of default) isn't that bad

Reply:
@Pennycook We have been giving this a thought and maybe returning a `void` pointer to align with C-like interfaces is not ideal. With the current proposal (missing an additional signature with alignment), users can pass `std::byte` as an `ElementType` and do type punning as they wish in case that's needed. Do you think we could leave this as is and only add a `void` pointer signature if users come up with an actual use-case for this?

This is an experimental feature anyway, so we could extend it in the future in cases like that.

Reply:
Yes, that's fine by me.